timApp/velp/velpgroups.py

```diff
@@ -359,12 +379,14 @@ def add_groups_to_selection_table(
         db.session.add(vgs)
 
 
-def process_selection_info(vgss: list[VelpGroupSelection] | list[VelpGroupDefaults]):
+def process_selection_info(
+    vgss: list[VelpGroupSelection] | list[VelpGroupDefaults],
+) -> dict[str, list | list[dict[str, int | bool]]]:
```

Comment:
Tämän tyypitystä pitäisi ehkä yksinkertaistaa. Yksi tapa on katsoa, miten tätä funktiota käytetään. Tämän funktion kanssa käy melkoinen onni, koska sen arvo palautetaan polussa `/<int:doc_id>/get_velp_group_personal_selections`:

https://github.com/TIM-JYU/TIM/blob/89346e440362cde1b40337e7f0f0ee2d56714658/timApp/velp/velp.py#L261

Eli voi katsoa, onko tuota tyypitetty selainkoodissa (hyvin todennäköisesti on, koska meillä on siellä vahva tyypitys). Näin näyttää olevan:

https://github.com/TIM-JYU/TIM/blob/a589686046537fa6912fac1f2e9765e5a7d97c19/timApp/static/scripts/tim/velp/velpSelection.ts#L197-L199

`IVelpGroupCollection` on muotoa

https://github.com/TIM-JYU/TIM/blob/4faf3572aeab376495982d647415991d3d8a733c/timApp/static/scripts/tim/velp/velptypes.ts#L193-L196

Toinen tapa olisi debuggerilla katsoa palautustyyppi ja tyypittää oikein. 

Tyypitys voidaan tehdä siis muutamalla tavalla:

* Antaa tyypiksi `dict`, koska palautetaan JSONia -> OK ja pitäisi mennä mypystä läpi, mutta tyypityksestä ei sitten ole mitään hyötyä
* Refaktoroidaan funktio käyttämään dataclasseja -> paras, mutta työläin, koska funktion logiikka pitäisi korjata
* Tehdään `TypedDict`, jolla voi määrittää tarkemman tyypityksen `dict`eille. Tässä tapauksessa `{id: number; selected?: boolean}` kääntyy Pythonissa muotoon
   ```py
   class VelpSelectionInfo(TypedDict):
     id: int
     selected: bool | None
   ```
   (paitsi onko `selected` itse asiassa `bool`, koska se ilmeisesti aina asetetaan `process_selection_info`:ssa joksikin arvoksi)
   
   Tällöin funktion palautustyyppi voi olla `dict[str, list[VelpSelectionInfo]]`, joka on tarkempi ja siistimpi.

TypedDictin miinus on toki se, että kyseessä on edelleen `dict` eikä dataclass, mutta ainakin näin saadaan alkuun järkevä tyypitys refaktoroimatta liikaa. Refaktorointia voi totta kai tehdä, mutta kaikkea koodia ei tarvitse kirjoittaa uusiksi heti.

Reply:
Tämä nyt bugikorjausten myötä vihdoin valmis.